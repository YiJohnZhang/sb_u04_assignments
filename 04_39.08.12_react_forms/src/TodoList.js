import React, {useState} from 'react';
import { v4 as uuidv4 } from 'uuid';

import Todo from './Todo';
import NewTodoForm from './NewTodoForm';


function TodoList(){

	const [todoList, setTodoList] = useState([]);

	function appendTodoEntry(todoItem){

		const newTodoEntry = {id:uuidv4(), ...todoItem};
		setTodoList([...todoList, newTodoEntry]);

	}

	function shiftTodoEntry(todoID){

		const newTodoList = todoList.filter((entry) => entry.id !== todoID);
		setTodoList(newTodoList);

	}

	function updateToDoEntry(todoID, todoData){


	}

	function toggleTodoStatus(todoID){

		const todoIDIndex = todoList.findIndex((entry) => entry.id === todoID);
		todoList[todoIDIndex].todoStatus = todoList[todoIDIndex].todoStatus === true ? false : true;
		setTodoList(todoList.map((entry) => entry));

	}

	return (

		<div>

			<NewTodoForm createToDoEntry={appendTodoEntry}/>
			<ul>

				{todoList.map((entry) => (
					<Todo
						key={entry.id}
						todoText={entry.todoText}
						completed={entry.todoStatus}
						deleteTodoEntry={() => shiftTodoEntry(entry.id)}
						editTodoEntry={() => updateToDoEntry}
						toggleTodoEntryStatus={() => toggleTodoStatus(entry.id)}
						/>
				))}
				
			</ul>

		</div>

	);
	
}

export default TodoList;

/* React Local Storage Notes:
	similar to vanilla: localStorage.setItem(...)
	then retrieve it by doing TodoList.defaultProps = localStorage.getItem(...)\

	React also allows to do custom hooks, i.e `useLocalStorage`
		also `useState` can dynamically set localStorage.
	https://blog.logrocket.com/using-localstorage-react-hooks/
*/