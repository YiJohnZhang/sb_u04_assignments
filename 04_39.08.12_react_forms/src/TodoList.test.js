import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import TodoList from './TodoList';

test('smoke test', () => {

	render(<TodoList />);

});

test('snapshot', () => {
	
	const {asFragment} = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();

});

test('fireEvent (form): fields update', () => {

	const { queryByTestId } = render(<TodoList />);
	const todoFormInput = queryByTestId('Todo-input_text');
	
	expect(Boolean(todoFormInput)).toBe(true);
	fireEvent.change(todoFormInput, {target: {value:'hello'}});
	expect(todoFormInput.value).toBe('hello');
		// yay

});

test('fireEvent (form): form submits and appends entry', () => {

	const { queryByTestId, queryByText } = render(<TodoList />);
	// console.log(queryByRole('textbox', {'description': 'entry...'}))
	const todoFormInput = queryByTestId('Todo-input_text');
	const todoFormCreate = queryByTestId('createTodo');
		// Consider standardizing testid naming conventions later
	expect(Boolean(todoFormInput)).toBe(true);
	expect(Boolean(todoFormCreate)).toBe(true);
	// console.log(Boolean(todoFormInput) == true);
	// console.log(Boolean(todoFormCreate) == true);
	
	fireEvent.change(todoFormInput, {target: {value:'hello'}});
	expect(todoFormInput.value).toBe('hello');

	fireEvent.click(todoFormCreate);
	expect(todoFormInput.value).toBe('');
	expect(queryByText("hello")).toBeInTheDocument();
		// this is guaranteed to be the new todo hello.

	
});

test('fireEvent (todo): delete todo', () => {

	const { queryByTestId, queryByText } = render(<TodoList />);
	const todoFormInput = queryByTestId('Todo-input_text');
	const todoFormCreate = queryByTestId('createTodo');
	expect(Boolean(todoFormInput)).toBe(true);
	expect(Boolean(todoFormCreate)).toBe(true);
	
	fireEvent.change(todoFormInput, {target: {value:'hello'}});
	fireEvent.click(todoFormCreate);
	expect(queryByText("hello")).toBeInTheDocument();
		// this is guaranteed to be the new todo hello.
	
	const deleteTodoEntryButton = queryByTestId('deleteTodo');
	fireEvent.click(deleteTodoEntryButton);
	expect(queryByText("hello")).not.toBeInTheDocument();

});