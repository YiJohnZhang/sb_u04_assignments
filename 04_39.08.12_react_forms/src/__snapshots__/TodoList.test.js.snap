// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="todoText"
      >
        <strong>
          Todo Entry
        </strong>
        : 
      </label>
      <input
        data-testid="Todo-input_text"
        name="todoText"
        placeholder="entry..."
        type="text"
        value=""
      />
      <button
        data-testid="createTodo"
      >
        New Item
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
